### PR TITLE
Hapi v16.1, Inert, nes, hapi-auth-jwt2, hapi-decorators.  New definitions for catbox, h2o2, mime-db, mimos, podium, shot, vision, hapi-auth-basic

### PR DESCRIPTION
**Pull request by AJamesPhillips**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://hapijs.com/api/16.1.1
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dslint/dt.json" }`. -- **not able to invest time to get node 7.x running in virtual env just to get linter running**

Concludes discussion on #15963 

Note this PR also contains fixes and updates to the Inert typings due to a previous mistake and changes to hapi typings that the Inert typings depend on.

### CODE REVIEW

**Comment by AJamesPhillips** [comment]
It's erroring due to "Expected file 'tests/connection/table.ts' to be named hapi-tests.ts".  Is it possible to merge this without having to rename all the test files.  I do not want to merge them into one monolithic file as this makes them significantly less useful and easy to keep updated.  The monolithic types file is already bad enough:  http://stackoverflow.com/questions/43276921/defining-module-and-plugins-with-core-module-having-multiple-files

**Comment by mmc41** [comment]
One question: Why does generate-response.ts contain this line ?
`const Promise = require('promise');`

It should not be necessary and it could break things.

**Comment by mmc41** [comment]
See also PR #15794 - there might be conflicts.

**Comment by jasonswearingen** [comment]
looks good, but the CI test failed, here are the details:

> Error: In /home/travis/build/DefinitelyTyped/DefinitelyTyped/types/hapi: Expected file 'tests/connection/table.ts' to be named hapi-tests.ts

**Comment by AJamesPhillips** [comment]
@jasonswearingen yeah seen :| See my comment a few back: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/16065#issuecomment-296393382
I hope we can merge without changing this. `tsc` runs correctly in hapi directory (catches type errors) and currently passes without error.

**Comment by AJamesPhillips** [comment]
Yeah, it should actually be `response?: Response | null;` as we have `"strictNullChecks": true` in tsconfig.  I was just being lazy as I had just fixed this example [from the hapi docs](https://hapijs.com/tutorials/cookies?lang=en_US):

    const server = new Hapi.Server();
    server.state('data', {ttl: null});

With the original .d.ts of: 

    export interface ServerStateCookieConfiguationObject {
        ttl?: number;
        ...
    }

It gives the error:

    Types of property 'ttl' are incompatible.
      Type 'null' is not assignable to type 'number | undefined'.

**Comment by AJamesPhillips** [comment]
@jasonswearingen this is a little more on it: https://github.com/Microsoft/TypeScript/issues/7426

**Comment by AJamesPhillips** [comment]
Actually sorry scrap all of that.  The docs state:

> the response object when set. The object can be modified but must not be assigned another object. To replace the response with another from within an extension point, use reply(response) to override with a different response. **Contains null when no response has been set** (e.g. when a request terminates prematurely when the client disconnects).

**Comment by AJamesPhillips** [comment]
@sheetalkamat it would be good to get this merged in or an indication from the team as to what needs to happen first.  If we need to rename / merge the tests then so be it.  I will try first to move them all to sub directories of their name with a script to undo it.  Would be a shame to spend effort on it though if it's something @DefinitelyTyped has imminent plans to fix/support.  Thanks for all the great work so far!

File: types/hapi/catbox/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,249 @@
+// TODO needs tests
+// TODO move to own definition
+// Type definitions for catbox 7.1
+// Project: https://github.com/hapijs/catbox
```
Comment (mhegazy) [inline types/hapi/catbox/index.d.ts]: these all seem to be valid node modules. they should live in top-level folders and not under `hapi`

**Comment by mhegazy** [comment]
@AJamesPhillips sorry for the delay. can you address the merge conflicts and the CI test failures, and we should be good to go

**Comment by mhegazy** [comment]
also looks like mimos, catbox, h2o2 are all node modules, so consider having them in top-level modules, instead of files under hapi.

**Comment by AJamesPhillips** [comment]
@mhegazy I'm certainly happy to:

* break out the definitions into their own to level
* Fix merge conflict (again :) )

Question remains over tests though.  Failure message was: `Expected file 'tests/connection/table.ts' to be named hapi-tests.ts.`  I feel tests are much more easily navigated/discoverable/maintainable at the monument due to their directory structure.  As I said to @sheetalkamat:

> If we need to rename / merge the tests then so be it. I will try first to move them all to sub directories of their name with a script to undo it. Would be a shame to spend effort on it though if it's something @DefinitelyTyped has imminent plans to fix/support.

Again thanks for the response.  Much appreciated.

**Comment by AJamesPhillips** [comment]
Also if you have any comment on what to do with the JSON typings.  Perhaps they should live in lib/browser/node typings some where?

`json/index.d.ts`

```typescript
/**
 * @see {@link https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#The_replacer_parameter}
 */
export type StringifyReplacer = ((key: string, value: any) => any) | (string | number)[] | undefined;

/**
 * Any value greater than 10 is truncated.
 */
export type StringifySpace = number | string;

export interface StringifyArguments {
    /** the replacer function or array. Defaults to no action. */
    replacer?: StringifyReplacer;
    /** number of spaces to indent nested object keys. Defaults to no indentation. */
    space?: StringifySpace;
}
```

`json/json-tests.ts`

```typescript
import * as JSON from './index';

var a: JSON.StringifyReplacer = function(key, value) {
    if (key === "do not include") {
        return undefined;
    }
    return value;
};
```

**Comment by jasonswearingen** [comment]
@AJamesPhillips  the replacer syntax is inline with the ```JSON.stringify()``` function definition, so it would need modification of ```lib.d.ts```

I don't actually think @mhegazy is requiring that all the "dependency modules" need to be split into their own DT top-level modules (it's nice to do for those mentioned of course)   but especially if you need to expand on something in ```lib.d.ts``` I *personally* think embeding as you did is fine.

**Comment by mhegazy** [comment]
> I don't actually think @mhegazy is requiring that all the "dependency modules" need to be split into their own DT top-level modules (it's nice to do for those mentioned of course) but especially if you need to expand on something in lib.d.ts I personally think embeding as you did is fine.

If the module is present on npm, then it should be in its own folder. otherwise, it should not.

**Comment by mhegazy** [comment]
> Question remains over tests though. Failure message was: Expected file 'tests/connection/table.ts' to be named hapi-tests.ts. I feel the tests are much more easily navigated/discoverable/maintainable at the moment due to their directory structure. As I said to @sheetalkamat:

the folder name should be `test` and not `tests`. the linter has special handling for a folder with that name. see example in https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/mithril/test

**Comment by AJamesPhillips** [comment]
@jasonswearingen ok great to know.  Could you point me towards where lib.d.ts entries are submitted please?  Is it on definitely typed?  (I can't see a `types/lib` directory).

@mhegazy is there any fix for the major version number?  I have put it back to 15 for the moment but I assume people's PRs before this one were for finding inaccuracies between it and v16 though from the [hapi v16.0 release notes](https://github.com/hapijs/hapi/issues/3398) the API only changed a little between 15 and 16.

**Comment by jasonswearingen** [comment]
@AJamesPhillips actually sorry I don't know where to submit changes to lib.d.ts :(  it's part of the typescript package so I am guessing it's something like https://github.com/Microsoft/TypeScript/blob/master/lib/lib.d.ts  but perhaps @mhegazy  can inform better on that.

**Comment by AJamesPhillips** [comment]
That makes sense.  Especially so as mhegazy said before: "If the module is present on npm, then it should be in its own folder. otherwise, it should not."  And thanks for the link.  Yes JSON is defined here:  https://github.com/Microsoft/TypeScript/blob/master/lib/lib.d.ts#L952-L979  good to see stringify arguments are the same as the definition here:  https://github.com/DefinitelyTyped/DefinitelyTyped/pull/16065/commits/ae60b8ca46d277d558d74afe2e03ab55f57d16b5#diff-074ce399e85a82d31a5349f977d389beR2562

**Comment by AJamesPhillips** [comment]
@NoHomey @warrenseymour I've been updating and done an overhaul of the hapi 16.1 typings.  In the process I (unfortunately) broke nes and hapi-auth-jwt2 typings.  I have fixed these but the changes are in a later commit you haven't been auto notified and had a chance to give any feedback on these.  So please feel free to point out errors or better ways of doing these.  Thanks!

**Comment by AJamesPhillips** [comment]
@kenhowardpdx @nycdotnet I've been updating and done an overhaul of the hapi 16.1 typings. In the process I (unfortunately) broke hapi-decorators and inert typings. I have fixed these but the changes are in a later commit you haven't been auto notified and had a chance to give any feedback on these. So please feel free to point out errors or better ways of doing these. Thanks!  (sorry I missed you when messaging NoHomey warrenseymour earlier)

**Comment by warrenseymour** [comment]
@AJamesPhillips Thanks, the `hapi-auth-jwt2` typings look good, although I've only read the diff.

Now that you've introduced the `PluginFunction` type, does that mean that https://github.com/DefinitelyTyped/DefinitelyTyped/pull/16065/files#diff-0e5a88f6b2a90cb6c0242c47fab96594R128 should be `PluginFunction<Options>`?

**Comment by AJamesPhillips** [comment]
@warrenseymour Hey!  Thanks for the (incredibly) speedy response 👍 .  It is very confusing as there are "Options" all over the place.  In the case of hapi-auth-jwt2 `Options` refers to those passed to `server.auth.strategy('jwt', 'jwt', options ...`.  The `Options` in `PluginFunction<Options>` refers to those passed to the [plugin function](https://hapijs.com/api/16.1.1#plugins) during registration.  So we're ok here as hapi-auth-jwt2 plugin doesn't accept any options.  Good question though, thanks.  Anything else you think might need changing please shout.

**Comment by warrenseymour** [comment]
@AJamesPhillips That certainly makes sense. Now, does this mean that `request.generateResponse()` now has valid typings? If it does, you've made me very happy.

**Comment by jasonswearingen** [comment]
looks like all checks have passed,  as this pull request touches a lot of typings, @mhegazy lets get this merged ASAP?

**Comment by kenhowardpdx** [comment]
@AJamesPhillips `hapi-decorators` LGTM.

**Comment by AJamesPhillips** [comment]
@warrenseymour yes `request.generateResponse()` is in.  Passes [this test](https://github.com/AJamesPhillips/DefinitelyTyped/blob/22d7c1dbcf8154db3578c5db5f2254ae3c39c3d8/types/hapi/test/request/generate-response.ts) at least. :)

File: types/hapi/index.d.ts [hunk 2]
```diff
@@ -1,2489 +1,2599 @@
...
-
- The framework comes with a few built-in handler types available by setting the route handler config to an object containing one of these keys.*/
-export interface IRouteHandlerConfig {
-	/** generates a static file endpoint for serving a single file. file can be set to:
-	 a relative or absolute file path string (relative paths are resolved based on the route files configuration).
-	 a function with the signature function(request) which returns the relative or absolute file path.
-	 an object with the following options */
-	file?: string | IRequestHandler<void> | IFileHandlerConfig;
-	/** directory - generates a directory endpoint for serving static content from a directory. Routes using the directory handler must include a path parameter at the end of the path string (e.g. /path/to/somewhere/{param} where the parameter name does not matter). The path parameter can use any of the parameter options (e.g. {param} for one level files only, {param?} for one level files or the directory root, {param*} for any level, or {param*3} for a specific level). If additional path parameters are present, they are ignored for the purpose of selecting the file system resource. The director
```
Comment (SimonSchick) [inline types/hapi/index.d.ts]: I don't think this field is mandatory, just updated typings and this is causing errors.

Comment (SimonSchick) [inline types/hapi/index.d.ts]: Shouldn't this be `ServerExtConfigurationObject[]`?
See: https://hapijs.com/api#route-options

Comment (SimonSchick) [inline types/hapi/index.d.ts]: `ServerExtConfigurationObject | ServerExtConfigurationObject[]` rather

**Comment by AJamesPhillips** [comment]
@aozgaa it was lost in the conversation above but:

> @mhegazy is there any fix for the major version number? I have put it back to 15 for the moment but I assume people's PRs before this one were for finding inaccuracies between it and v16. If it has to stay as v15 for the moment then it shouldn't be too big a deal as the hapi v16.0 release notes show the API only changed a little between 15 and 16.

**Comment by mmc41** [comment]
@AJamesPhillips Thanks for all the good work. Not sure if I should point out regression errors in new issues or just here but unfortunately, there seem to an issue with joi validation if using the new types. The definition of params and query does not anymore allow me to specify detailed constrains for header parameters and query parameters. I get a compilation error if I try code like the example below which worked with prior versions of hapi types:

```
server.route({ 
        method: 'GET',
        path: ...,
        handler: function (request: hapi.Request, reply: hapi.ReplyNoContinue): void {
         .....  
        },
        config: {
            validate: {
                headers: true,
                params: { // No longer allowed??
                    id: Joi.string()
                }
                query: { // No longer allowed??
                    providerId: Joi.string()
                }
            }
        }
    });
```

**Comment by AJamesPhillips** [comment]
Hey @mmc41, was going to say you can post here or in a new issue but actually it's best to create a new issue, reference this PR with #16065 and @ mention myself or jasonswearingen  if you had a question otherwise we might miss it.  Preferable to all of that: fire up a PR and go for it 👍  😄 

Re the regression above, you got a link to that example in the docs or you just know this works?

**Comment by mmc41** [comment]
Fixed in newly released 16.1.2 :-) @AJamesPhillips Thanks

**Comment by AJamesPhillips** [comment]
@garthk yes it would have been preferable to have a major version number update however hapi was already at v16, the typings were at v16 (though in practice were very out of date since jasonswearingen's excellent major effort at v12).

You'll need to use v15 if you want the older typings, apparently [it's not possible to use 16 and 16.1 as the question has gone unanswered](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/16065#issuecomment-301699996) and [the `@types` publishing code shows it's not possible at the moment](https://github.com/Microsoft/types-publisher/blob/cfd90020c361c7a90ea94c333d983cb605eed4fe/src/lib/definition-parser.ts#L28-L30).

Having said that it still wouldn't have helped in your case as you'd like a semver-major.  This unfortunately is not how the `@types` system works at the moment or really can work in any sane way in the future.  It would be the road to madness to publish this major revision of hapi 16.1 typings as `@types/hapi` 17.0.  Sorry to hear your build broke, it is annoying.  I personally lock all typings (and all packages) to only patch updates if it's a noncritical build and locked entirely if it's a critical build.  Definitely for typings this is a good strategy.

**Comment by AJamesPhillips** [comment]
I missed part of your post before @garthk , what's the "Java taste" part?

**Pull request merged**